### PR TITLE
dragonnet implementation + semi-synthetic dataset benchmark comparison

### PR DESCRIPTION
**Pull request by yungmsh**
- added implementation of dragonnet in `inference.nn.dragonnet` (code modified from https://github.com/claudiashi57/dragonnet)
- added example notebook to compare performance against causalml's meta learners (using a subset of IHDP semi-synthetic data - the data they used in their paper). Based on the paper (Page 7 in https://arxiv.org/pdf/1906.02120.pdf), they used "1000 realizations from the NPCI package" (which is an R package https://github.com/vdorie/npci). In the dragonnet repo, I found 50 csvs, so this notebook compares performance for those 50 datasets. I did, however, run into some keras/tensorflow errors like "Batch X: Invalid loss, terminating training" on about 8 of the 50 datasets - need to look into what's causing that.

### CODE REVIEW

File: causalml/inference/nn/dragonnet.py [hunk 1]
```diff
@@ -0,0 +1,255 @@
...
+from keras.layers import Flatten
+from keras.layers import Input
+import tensorflow as tf
+import keras.backend as K
+from keras.engine.topology import Layer
+from keras.metrics import binary_accuracy
+from keras.layers import Input, Dense, Concatenate, BatchNormalization, Dropout
+from keras import regularizers
+from keras.optimizers import rmsprop, SGD, Adam
+from keras.callbacks import EarlyStopping, ModelCheckpoint, TensorBoard, ReduceLROnPlateau, TerminateOnNaN
```
Comment (jeongyoonlee) [inline causalml/inference/nn/dragonnet.py]: Let's clean up imports by removing unused imports: `pandas`, `matplotlib`, `scipy.special`, `StandardScaler`, `train_test_split`, `Conv2D`, `MaxPooling2D`, `Activation`, `Lambda`, `Flatten`, `Input`, `Dense`, `BatchNormalization`, `rmsprop`, `ModelCheckpoint`, `TensorBoard`.

File: causalml/inference/nn/dragonnet.py [hunk 2]
```diff
@@ -0,0 +1,255 @@
...
+
+def make_dragonnet(input_dim, reg_l2, neurons_per_layer=200):
+    """
+    Neural net predictive model. The dragon has three heads.
+    :param input_dim:
+    :param reg:
+    :return:
+    """
+    t_l1 = 0.
+    t_l2 = reg_l2
```
Comment (jeongyoonlee) [inline causalml/inference/nn/dragonnet.py]: Unused variables.

File: causalml/inference/nn/dragonnet.py [hunk 3]
```diff
@@ -0,0 +1,255 @@
...
+
+def train_dragon(t, y_unscaled, x, targeted_regularization=True, output_dir='',
+                              knob_loss=dragonnet_loss_binarycross, ratio=1., dragon='', val_split=0.2,
+                              neurons_per_layer=50, batch_size=64, epochs=30, learning_rate=1e-3, verbose=True):
+    verbose = verbose
+    # y_scaler = StandardScaler().fit(y_unscaled)
+    # y = y_scaler.transform(y_unscaled)
+    y = np.hstack((y_unscaled, t))
+    train_outputs = []
+    test_outputs = []
```
Comment (jeongyoonlee) [inline causalml/inference/nn/dragonnet.py]: Unused variables.

File: causalml/inference/nn/dragonnet.py [hunk 4]
```diff
@@ -0,0 +1,255 @@
...
+        loss = knob_loss
+
+    # for reporducing the IHDP experimemt
+#     train_index, test_index = train_test_split(np.arange(x.shape[0]), test_size=0.2, random_state=1)
+
+#     x_train, x_test = x[train_index], x[test_index]
+#     y_train, y_test = y[train_index], y[test_index]
+#     t_train, t_test = t[train_index], t[test_index]
+
+#     yt_train = np.concatenate([y_train, t_train], 1)
```
Comment (jeongyoonlee) [inline causalml/inference/nn/dragonnet.py]: Remove commented code.

File: causalml/inference/nn/dragonnet.py [hunk 5]
```diff
@@ -0,0 +1,255 @@
...
+    return dragonnet
+
+    # yt_hat_test = dragonnet.predict(x_test)
+    # yt_hat_train = dragonnet.predict(x_train)
+
+    # test_outputs += [_split_output(yt_hat_test, t_test, y_test, y_scaler, x_test, test_index)]
+    # train_outputs += [_split_output(yt_hat_train, t_train, y_train, y_scaler, x_train, train_index)]
+    # K.clear_session()
+
+    # return test_outputs, train_outputs
```
Comment (jeongyoonlee) [inline causalml/inference/nn/dragonnet.py]: Remove commented code.

File: examples/ihdp_dataset_dragonnet_vs_metalearners.ipynb [hunk 6]
```diff
@@ -0,0 +1,1767 @@
...
+       "    <tr>\n",
+       "      <th>T</th>\n",
+       "      <td>4.698523</td>\n",
+       "      <td>1.132001</td>\n",
+       "      <td>0.580085</td>\n",
+       "    </tr>\n",
+       "    <tr>\n",
+       "      <th>X</th>\n",
+       "      <td>4.534673</td>\n",
+       "      <td>1.2382
```
Comment (jeongyoonlee) [inline examples/ihdp_dataset_dragonnet_vs_metalearners.ipynb]: Unused import. Let's remove it.

File: examples/ihdp_dataset_dragonnet_vs_metalearners.ipynb [hunk 7]
```diff
@@ -0,0 +1,1767 @@
...
+    }
+   },
+   "outputs": [],
+   "source": [
+    "df_result = pd.DataFrame([s_ate, t_ate, x_ate, r_ate, dragon_ate, tau.mean()],\n",
+    "                     index=['S','T','X','R','dragonnet','actual'], columns=['ATE'])\n",
+    "df_result['MAE'] = [mae(t,p) for t,p in zip([s_ite, t_ite, x_ite, r_ite, dragon_ite],\n",
+    "                                         [tau.values.reshape(-1,1)]*5 )\n",
+    "                ] + [None]\n",
+    "df_result['AUUC'] = auuc_score(df_preds)"
```
Comment (jeongyoonlee) [inline examples/ihdp_dataset_dragonnet_vs_metalearners.ipynb]: `auuc_score()` is not defined.

File: examples/ihdp_dataset_dragonnet_vs_metalearners.ipynb (hunk 6 above)
Comment (jeongyoonlee) [inline examples/ihdp_dataset_dragonnet_vs_metalearners.ipynb]: Is it possible to download the file in the notebook? If so, let's add code to download the IHDP data set. If not, let's gzip and add one of CSV files to the repo and make it available. 

Also let's remove your personal paths in the notebook.

File: causalml/inference/nn/dragonnet.py [hunk 8]
```diff
@@ -0,0 +1,209 @@
...
+            treatment (np.array or pd.Series): a treatment vector
+            y (np.array or pd.Series): an outcome vector
+        """
+        X, treatment, y = convert_pd_to_np(X, treatment, y)
+
+        y = np.hstack((y.reshape(-1, 1), treatment.reshape(-1, 1)))
+
+        self.dragonnet = self.make_dragonnet(X.shape[1])
+
+        metrics = [regression_loss, binary_classification_loss, treatment_accuracy, track_epsilon]
```
Comment (jeongyoonlee) [inline causalml/inference/nn/dragonnet.py]: What is `epsilon` here? How is it different from the propensity?

File: causalml/inference/nn/utils.py [hunk 9]
```diff
@@ -0,0 +1,97 @@
+import tensorflow as tf
+import keras.backend as K
+from keras.engine.topology import Layer
+from keras.metrics import binary_accuracy
+
+
+def binary_classification_loss(concat_true, concat_pred):
```
Comment (jeongyoonlee) [inline causalml/inference/nn/utils.py]: Could you add docstrings to functions here?

Comment (yungmsh) [reply causalml/inference/nn/utils.py]: docstrings added in latest push

File: causalml/inference/nn/utils.py [hunk 10]
```diff
@@ -0,0 +1,161 @@
...
+        epsilons = concat_pred[:, 3]
+        t_pred = (t_pred + 0.01) / 1.02
+        # t_pred = tf.clip_by_value(t_pred,0.01, 0.99,name='t_pred')
+
+        y_pred = t_true * y1_pred + (1 - t_true) * y0_pred
+
+        h = t_true / t_pred - (1 - t_true) / (1 - t_pred)
+
+        y_pert = y_pred + epsilons * h
+        targeted_regularization = tf.reduce_sum(tf.square(y_true - y_pert))
```
Comment (yungmsh) [inline causalml/inference/nn/utils.py]: @jeongyoonlee epsilon is used here. y_pert is an adjusted version of y_pred (adjusted by epsilon times h, which is the targeted regularization part). So epsilon's just another hyperparameter to be tuned it looks like.

File: causalml/inference/nn/dragonnet.py (hunk 8 above)
Comment (yungmsh) [reply causalml/inference/nn/dragonnet.py]: see comment here: https://github.com/uber/causalml/pull/128/files#r373743965
